File: backend/app/schemas/conflict.py
"""
backend/app/schemas/conflict.py
Pydantic v2 schemas untuk conflict detection API.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class ConflictLogResponse(BaseModel):
    """Schema untuk satu entri ConflictLog."""

    id: uuid.UUID
    sesi_id: uuid.UUID
    jenis: str
    severity: str
    assignment_ids: List[uuid.UUID]
    pesan: str
    detail: Optional[Dict[str, Any]] = None
    checked_at: datetime
    is_resolved: bool

    model_config = {"from_attributes": True}


class ConflictCheckSummary(BaseModel):
    """Ringkasan hasil check-conflicts: jumlah ERROR dan WARNING."""

    sesi_id: uuid.UUID
    total_error: int
    total_warning: int
    total: int
    conflicts: List[ConflictLogResponse]


class ConflictListResponse(BaseModel):
    """Daftar konflik dari conflict_log terbaru dengan pagination."""

    items: List[ConflictLogResponse]
    total: int


class ConflictResolveResponse(BaseModel):
    """Response setelah menandai konflik sebagai resolved."""

    id: uuid.UUID
    is_resolved: bool
    pesan: str

    model_config = {"from_attributes": True}

File: frontend/src/api/conflict.ts
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface ConflictLog {
  id: string
  sesi_id: string
  jenis: string
  severity: 'ERROR' | 'WARNING'
  assignment_ids: string[]
  pesan: string
  detail: Record<string, unknown> | null
  checked_at: string
  is_resolved: boolean
}

export interface ConflictListParams {
  jenis?: string
  severity?: 'ERROR' | 'WARNING'
}

export interface ConflictSummary {
  total_error: number
  total_warning: number
}

// ─── API Functions ────────────────────────────────────────────────────────────

export async function getConflicts(
  sesiId: string,
  params?: ConflictListParams
): Promise<ConflictLog[]> {
  const res = await apiClient.get<ConflictLog[]>(`/sesi/${sesiId}/conflicts`, { params })
  return res.data
}

export async function checkConflicts(sesiId: string): Promise<ConflictSummary> {
  const res = await apiClient.post<ConflictSummary>(`/sesi/${sesiId}/check-conflicts`)
  return res.data
}

export async function resolveConflict(
  sesiId: string,
  conflictId: string
): Promise<ConflictLog> {
  const res = await apiClient.patch<ConflictLog>(
    `/sesi/${sesiId}/conflicts/${conflictId}/resolve`
  )
  return res.data
}
